### ISSUE TITLE
Can Weak be null?

### ISSUE THREAD

**Issue by est31** [rust-lang/rust#114517]
There is a discrepancy between the docs of `Weak` and its implementation, originally found by @Urgau in https://github.com/rust-lang/rust/pull/114494#discussion_r1285009958 .

On one hand, the [docs of Weak::as_ptr](https://doc.rust-lang.org/stable/std/rc/struct.Weak.html#method.as_ptr) state:

> The pointer is valid only if there are some strong references. The pointer may be dangling, unaligned or even [null](https://doc.rust-lang.org/stable/std/ptr/fn.null.html) otherwise.

On the other hand, the implementation of `Weak` uses `NonNull`:

https://github.com/rust-lang/rust/blob/1cabb8ed2301faf193132e71ca5b10f82913e48a/library/alloc/src/rc.rs#L2681-L2693

From what I know, it is okay for `NonNull` pointers to be dangling or unaligned, but, staying true to their name, they are [never allowed to be null pointers](https://doc.rust-lang.org/stable/std/ptr/struct.NonNull.html#safety-2).

I think this discrepancy should be resolved, but I wonder, which way? Should the docs be adjusted, or the implementation?

**Comment by est31** [comment]
Some research on the history: `Weak::as_ptr` was introduced by #60766 (under a different name), tracking issue #60728. That PR already included the note that the pointer can't be null, even though already back then, `Weak` was using a `NonNull` internally.

**Comment by Zalathar** [comment]
I don't think there's a discrepancy here?

The documentation says that the method isn't guaranteed to return a non-null pointer.

The current implementation *happens* to never return a null pointer due to an internal detail (using `NonNull`), but that detail isn't a documented guarantee, so callers can't rely on it.

(For example, changing the implementation to return a null pointer in some circumstances would not be a breaking change, because the documented behaviour allows that possibility, even though the current implementation doesn't exercise it.)

**Comment by RalfJung** [comment]
Yeah, the key point is that `from_raw` does not allow null pointers.

**Comment by est31** [comment]
>  but that detail isn't a documented guarantee, so callers can't rely on it.

Good point, makes sense.

> `from_raw` does not allow null pointers.

This is not written anywhere in the docs of `from_raw`. It only says that the weak has to come from `into_raw`. The obvious alternatives like `as_ptr` plus forget (which `from_raw` is a wrapper of), or `into_raw_and_alloc` are missing. `into_raw` mentions nowhere that it can return a null pointer either.

I think there is a small, but non zero chance that when people read that the output of `as_ptr` can be a null pointer, they could assume that `from_raw` can take the input of `as_ptr` as well, and therefore, also a null pointer.

I'll make a PR to improve the docs.

**Comment by est31** [comment]
PR filed: #114525
